File: haskell/luhn/Luhn.hs
module Luhn (checkDigit, addends, checksum, isValid, create) where
-- module Luhn where

import Data.List(unfoldr)

addends :: Integer -> [Integer]
addends number = reverse . map fst $ unfoldr doAddends (number, False)
    where
        doAddends (n, double)
            | n == 0 = Nothing
            | not double = justPair $ n `mod` 10
            | n `mod` 10 >= 5 = justPair $ 1 + (2*n) `mod` 10
            | otherwise = justPair $ (2*n) `mod` 10
            where
                double'' = not double
                next = (n `div` 10, double'')
                justPair x = Just ((x, double), next)


create :: Integer -> Integer
create input = deslocInput + ((9 * addendsSum) `mod` 10)
    where
        deslocInput = 10 * input
        addendsSum = checksum deslocInput

checkDigit :: Integer -> Integer
checkDigit = (`mod` 10)

checksum :: Integer -> Integer
checksum = checkDigit . sum . addends

isValid :: Integer -> Bool
isValid = (0 ==) . checksum

File: python/luhn/luhn.py
class Luhn(object):
    def __init__(self, number):
        self.number = number

    def addends(self):
        result = []
        double = False
        number = self.number

        while number > 0:
            digit = number % 10
            if double:
                digit *= 2
                digit = digit % 10 + 1 if digit > 9 else digit
            double = not double

            result.append(digit)
            number = int(number / 10)

        return result

    def checksum(self):
        return sum(self.addends()) % 10

    def is_valid(self):
        return self.checksum() == 0

    def create(number):
        number *= 10
        luhn_number = Luhn(number)

        check_digit = 9 * luhn_number.checksum() % 10
        return number + check_digit
